### PR TITLE
Feature/auto casting function arguments

### LINKED ISSUE #12: Implement Type Promotion
See [https://www.w3.org/TR/xpath-31/#promotion](https://www.w3.org/TR/xpath-31/#promotion), mostly relevant for the arithmetic operators.

### PR DESCRIPTION
**Pull request by jitsedesmet**
This pull request aims to provide an implementation on auto casting functions as talked about in #102 . 
This pull request should *NOT* be merged before the scheme talked about in #102 is added.

There are a few questions to be answered before merging this PR too, those questions should be marked with a `TODO`.

Edit:
We can close a few issues with this PR: #13, #94 and #103.
When merging we open issues about: cast decimal so we can close #12 , questions surrounding `arithmeticWidening`, revision of https://github.com/comunica/sparqlee/pull/103#discussion_r687755303 (I think this issue is mainly for @wschella) and an issue surrounding caching and open world types.
The PR also makes implementation of #87 a lot easier.

### CODE REVIEW

File: lib/functions/OverloadTree.ts
Comment (rubensworks) [inline lib/functions/OverloadTree.ts:7]: OverrideType

File: lib/functions/OverloadTree.ts
Comment (rubensworks) [inline lib/functions/OverloadTree.ts:36]: I think that's correct indeed.

**Review by rubensworks — COMMENTED** [review]
Given the size of this PR, I'm a bit scared of the lack of tests related to this.
We did indeed have tests already that will now cover part of this implementation, but looking at the coveralls output, it seems like coverage is not complete.

Therefore, I would recommend to add more detailed unit tests to the parts of the code you touched. Should be straightforward to do, now that this code is still fresh.

**Comment by jitsedesmet** [comment]
> Therefore, I would recommend to add more detailed unit tests to the parts of the code you touched. Should be straightforward to do, now that this code is still fresh.

Do you think about changing the test directory structure at the same time? Like talked about in #34 ? So we can create a dir for unit tests as well?

File: lib/functions/OverloadTree.ts
Comment (jitsedesmet) [reply lib/functions/OverloadTree.ts:36]: Yes, after reading https://github.com/comunica/sparqlee/issues/102#issuecomment-896876623 and in particular https://www.w3.org/TR/xpath-31/#promotion it seems like that is the right behavior. I will add the reference to the code.

**Comment by rubensworks** [comment]
> Do you think about changing the test directory structure at the same time? Like talked about in #34 ? So we can create a dir for unit tests as well?

Oh, I thought we did that one already 😅 

I think that's quite urgent, yes. But I would do it in a separate PR.

File: lib/functions/OverloadTree.ts
Comment (jitsedesmet) [reply lib/functions/OverloadTree.ts:36]: Taking this into account I think we should also allow the numeric types in this PR. https://github.com/comunica/sparqlee/issues/102#issuecomment-897103472

**Comment by wschella** [comment]
I can review the PR this afternoon. I think changing the test folder structure can be done & and merged quickly. Then this PR can be rebased on it  (since it will be a significant one, and I do really propose fixing everything mentioned in #102 at once here, it's too closely related). We should indeed strive to not decrease test coverage here, but everything should be able to be tested in theory with some entries in a `TestTable`  somewhere.

File: lib/functions/OverloadTree.ts
Comment (jitsedesmet) [reply lib/functions/OverloadTree.ts:36]: Seems like it is not correct. Please take a look at https://github.com/comunica/sparqlee/issues/102#issuecomment-897427439 .

File: lib/functions/OverloadTree.ts
Comment (jitsedesmet) [inline lib/functions/OverloadTree.ts:116]: This code should allow type promotion as talked about in https://www.w3.org/TR/xpath-31/#promotion . At least, how I understand it. If this is wrong please let me know.

Comment (rubensworks) [reply lib/functions/OverloadTree.ts:116]: At first glance it looks like it could work.
But we'll now for certain once we have tests covering these cases.

Comment (jitsedesmet) [reply lib/functions/OverloadTree.ts:116]: Yes, I will add the scheme first and then start writing the tests. With 'wrong' I meant the interpretation of 'promotion'. :)

File: README.md
Comment (jitsedesmet) [inline README.md:338]: @wschella, the things that I removed here, what do they mean? 
Anyway, I should take a look at this!

File: lib/functions/OverloadTree.ts
Comment (wschella) [inline lib/functions/OverloadTree.ts:25]:
```suggestion
   * Get the implementation for the types that exactly match @param args .
```

File: lib/functions/RegularFunctions.ts
Comment (wschella) [inline lib/functions/RegularFunctions.ts:341]: See above

File: lib/util/Consts.ts
Comment (wschella) [inline lib/util/Consts.ts:80]: We should be able to get rid of this function in it's entirety (see discussion on type-attribute). Or am I missing something?

File: lib/util/TypeHandling.ts
Comment (wschella) [inline lib/util/TypeHandling.ts:5]:
```suggestion
export type OverrideType = LiteralTypes | 'term';

/**
```

File: test/integration/functions/onStrings.test.ts
Comment (wschella) [inline test/integration/functions/onStrings.test.ts:63]:
```suggestion
      "type promotion"^^xsd:anyURI 2 4 = "ype"
```

File: type-scheme.svg
Comment (wschella) [inline type-scheme.svg:1]: We should move this to some `docs/` folder.

**Review by wschella — CHANGES_REQUESTED** [review]
Very nice work, very important, and cleans things up nicely. 

I also think we we can make it a bit cleaner still: see comments and suggested changes.

File: lib/Transformation.ts
Comment (jitsedesmet) [reply lib/Transformation.ts:116]: You're right, keen eye!

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:11]: Do we have one for this? I'd need to check it. I thought you suggested creating this type for some cleaner code?

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:73]: Awesome, removing that would provided some even cleaner code. Personally I would go for `dataType` because I think `type` feels kind of empty. (It doesn't mean much)

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:144]: Right, about this, what doe it mean? It's kind of vague for an outsider, could you elaborate on it? Then I will also have a clearer view on what exactly to do with it.

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:148]: Yes, that's not clear at all. I wanted to have some way of telling this type was a `SPARQL_NUMERIC` but I ended up checking this in the constructor so it can go.

File: lib/functions/OverloadTree.ts
Comment (jitsedesmet) [reply lib/functions/OverloadTree.ts:90]: I would want to say 'to cast something in the past tense' but I have no idea if it is spelled correctly.

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:80]: If by `this function` you mean `decatogorize`, it's gone?

File: test/integration/functions/onStrings.test.ts
Comment (jitsedesmet) [reply test/integration/functions/onStrings.test.ts:23]: What do you mean?

File: lib/util/Consts.ts
Comment (rubensworks) [reply lib/util/Consts.ts:11]: Yes, this definitely needs some kind of description, because this doesn't appear to be part of the SPARQL/RDF spec (?).

File: test/integration/functions/onStrings.test.ts
Comment (rubensworks) [reply test/integration/functions/onStrings.test.ts:23]: In RDF, strings with a language tag can not have a datatype, and vice-versa.

File: type-scheme.svg
Comment (rubensworks) [reply type-scheme.svg:1]: Nice figure btw :-)

File: lib/util/Consts.ts
Comment (wschella) [reply lib/util/Consts.ts:11]: I did suggest it indeed, and it is present in the spec, but it is hidden a bit (can't put all the details up front ofc).

The first paragraphs of https://www.w3.org/TR/sparql11-query/#func-strings provide some more context. There is a separate section on argument compatibility that is relevant for some other issues as well. Let's summarize this together with the link and add it to the code.

File: lib/functions/OverloadTree.ts
Comment (wschella) [reply lib/functions/OverloadTree.ts]: A special thing to consider here is that subtype substitution and type promotion can work at the same time. There certainly should be tests for that, as I'm not sure the code handles this like it is (but it's also likely that I'm just missing something).

File: README.md
Comment (wschella) [reply README.md:338]: Resolved in suggestion.

File: lib/util/Consts.ts
Comment (wschella) [reply lib/util/Consts.ts:16]: `const enum` is a TS feature https://www.typescriptlang.org/docs/handbook/enums.html#const-enums that resolves some enum handling at compile time vs runtime. It might be worth to use these for speed, it might not be for other reasons, but we should at least write down why.

File: lib/expressions/Term.ts
Comment (wschella) [reply lib/expressions/Term.ts:73]: Agreed.

File: lib/expressions/Term.ts
Comment (wschella) [reply lib/expressions/Term.ts:144]: The SPARQL spec defines a way to format floats, the JS spec does the same. These aren't necessarily compatible. I can't recall why this is just commented out (it's just adds a trailing `.0` if the number is round), but it might be hint we don't have sufficient tests yet that the way we format things agrees with the SPARQL spec (because we use JS defaults).

File: lib/expressions/Term.ts
Comment (wschella) [reply lib/expressions/Term.ts:121]: The problem here is mostly runtime overhead I think. These are a lot of checks we need to for each number, which is a very common datatype.
If we have a subtypesFor, we can create a static map matching each concrete subtype to a concrete formatter a bit faster. Not sure if worth, just a gut feeling.

File: lib/functions/OverloadTree.ts
Comment (wschella) [reply lib/functions/OverloadTree.ts:90]: it's just `if it's cast first` then

File: README.md
Comment (jitsedesmet) [reply README.md:338]: I unresolved this since It's still imported for me to check the functions and I'll use this as a reminder :)

File: lib/Transformation.ts
Comment (jitsedesmet) [inline lib/Transformation.ts:86]: I was wondering if we could get some kind of reference on why we do this? It actually reminds me a bit of #14 ? Anyway, it's peculiar and I think we should write down what happens here.

Comment (jitsedesmet) [reply lib/Transformation.ts:86]: Same thing for the cases of null, undefined and '' underneath this.

File: lib/expressions/Term.ts
Comment (jitsedesmet) [inline lib/expressions/Term.ts:104]: Am I wrong to think we want to make strValue mandatory? If you give a literal to sparqlee and ask it back immediately. Shouldn't we get the same thing? This is particularly interesting for all types date related. If you where to give a date without a time zone, you wouldn't want to get one back with a arbitrary timezone? Wouldn't it pay of to generalize this?

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:11]: > take a string literal as an argument and accept a simple literal, a plain literal with language tag, or a literal with datatype xsd:string
This means a simple literal. Whatever that means, is also a sparql_stringly. We should make sure this is True as well.

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:16]: I'm thinking we don't want this? We probably want to add an extra option to the evaluatorConfig called `additionalExtentionTable` or smth. If we were to add this we need the runtime values. 
It's possible I don't understand const enum  enough but I think it would remove the runtime values.

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:144]: I agree, that's also a reason I commented https://github.com/comunica/sparqlee/pull/103#discussion_r688370200 .

Comment (jitsedesmet) [reply lib/expressions/Term.ts:144]: We might need to think about creating js Classses representing the various datatypes? I don't know if this would break things, if it would be a lot of work or if it's even possible. I do think it can be interesting for formatting purposes. In addition, our TypeUrl currently only extend 1 thing. That means we could map this perfectly to js class extension? Wouldn't that be great?

File: lib/Transformation.ts
Comment (wschella) [reply lib/Transformation.ts:86]: You mean the fact that if there's no datatype this is considered a string?

Comment (jitsedesmet) [reply lib/Transformation.ts:86]: Yes.
Also in https://www.w3.org/TR/sparql11-query/#func-lang I found
> Note that the RDF data model does not include literals with an empty language tag.

Does this have anything to do with this?

Comment (wschella) [reply lib/Transformation.ts:86]: It's this one again: https://www.w3.org/TR/sparql11-query/#func-string.

Comment (wschella) [reply lib/Transformation.ts:86]: Or at least largely. I don't think there are any functions that take an xsd:string but not a simple literal (no datatype).

Comment (wschella) [reply lib/Transformation.ts:86]: I think there has been some discussion in IDLab regarding this as well, and possible spec incompatibilities. Maybe @rubensworks has some better references.

Comment (wschella) [reply lib/Transformation.ts:86]: See also https://www.w3.org/TR/2004/REC-rdf-concepts-20040210/#dfn-plain-literal .

The distinction between simple literal and plain literal has never been clear to be, as plain literals are simple literals but with an (optional) language tag. So a plain one without a language tag is just a simple one?

The key part is simple, explicitly typed, and language fall under "string literal" (i.e. stringly in our case), and there are IIRC functions that take an explicitly typed xsd:string but not a simple (no datatype, no language) one.

There are functions that take xsd:string + simple, but don't take languageStrings.

Comment (rubensworks) [reply lib/Transformation.ts:86]: This issue should contain all the answers :-) https://github.com/w3c/sparql-12/issues/112

File: lib/util/Consts.ts
Comment (wschella) [reply lib/util/Consts.ts:11]: This true because we consider simple literal (no datatype) to be xsd:string (see also https://github.com/comunica/sparqlee/pull/103#discussion_r688454246).

File: lib/util/Consts.ts
Comment (wschella) [reply lib/util/Consts.ts:16]: This seems likely.

File: lib/util/Consts.ts
Comment (wschella) [reply lib/util/Consts.ts:80]: Something weird happened, I meant the `type(typeString: string): LiteralTypes` function.

File: lib/Transformation.ts
Comment (wschella) [reply lib/Transformation.ts:86]: Ah exactly. And for reference, we already do RDF 1.1 behavior. If there is no datatype, we make it xsd:string. (except when there is a language tag ofc)

File: lib/expressions/Term.ts
Comment (wschella) [reply lib/expressions/Term.ts:104]: This is optional because we don't format prematurely when calculating, as the value could be changed many times, and we only care for formatting when exporting.

If you simply pass a literal from somewhere else and request that, strValue should be populated, and will be returned. See the `str` function. 

Anyway, this is only for performance, you need to have correct formatting in any case.

Datetime is an exception because there are no functions that manipulate the (typed) value https://www.w3.org/TR/sparql11-query/#func-date-time .

Might be worth adding this to some docs.

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:121]: Those are indeed a lot of checks for every number. would it help to make a function that would do these checks only when needed in `str()` ?

Comment (wschella) [reply lib/expressions/Term.ts:121]: It's better. But I still think a static map is the ideal approach here. Your choice ofc. You can also split the numeric class into the primitives (float, decimal, double), but that won't solve much.

You need either (derived type) -> (primitive type) -> (formatter), or (derived type) -> (formatter). There's plenty of options for building one I think.

File: test/integration/functions/onStrings.test.ts
Comment (jitsedesmet) [reply test/integration/functions/onStrings.test.ts:23]: Oh I see, alright. Thanks

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:11]: Since I stumbled upon this problem like 5 times now. Should we put it in the readme or would that be overkill?

Comment (jitsedesmet) [reply lib/util/Consts.ts:11]: I'll just write it very clearly in the places I had issues with it. If I put it in the readme the chance is low someone will read it.

Comment (wschella) [reply lib/util/Consts.ts:11]: I don't think it's necessary. Maybe mentioning simple literals don't exists because we make them xsd:string is worth it.

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:16]: Okey, Should I keep the original `TODO` or can I remove it?

Comment (wschella) [reply lib/util/Consts.ts:16]: As you see fit.

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:144]: @rubensworks left me this: https://github.com/w3c/rdf-tests/issues/58  and that points to #28 . I think this refers to the code here? That would mean we can just remove the comment and there is no reason to handle it?

Comment (wschella) [reply lib/expressions/Term.ts:144]: Makes sense, let's clean it up.

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:121]: I don't think we can do a static map. Especially if we move sparqlee to have an open world mentality. Having a list of all subtypes would not be possible I think. Or just not scalable.

And like you said, splitting this class will not solve much. In that case we will need to do the check at construction like it is now. I'm sorry, I'll need to disappoint you. I will create a `specificFormattrCreator`.

Comment (wschella) [reply lib/expressions/Term.ts:121]: I don't think there is a scenario where the static maps don't scales but these checks do?

You are overestimating the likelihood of people providing _even more_ numeric subtypes, and sacrificing performance now for an unlikely scenario in a future world where sparqlee might be open world.

Even then, when it is open world, it would be preferable to have a static map, and a fallback mechanism for when the concrete type is not present.

File: lib/Transformation.ts
Comment (jitsedesmet) [reply lib/Transformation.ts:86]: Alright, thank you for this reference. Now I will add this reference to the code. Should I add a counter as well? :p (jk)

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:104]: Alright, thanks for the clarification.

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:16]: I think we can remove it.

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:121]: Maybe I just don't see it I'll make an issue about this and leave it with a creator?

Comment (wschella) [reply lib/expressions/Term.ts:121]: That's okay :)

File: lib/functions/OverloadTree.ts
Comment (jitsedesmet) [reply lib/functions/OverloadTree.ts]: I pretty sure this wouldn't be a problem since promotion is always from a non super type. To another type. And of course like it is now, having a restriction of anyURI provided to a function expecting an xsd_string, the restriction will be substituted to anyUri and will then be promoted to xsd_string.

**Comment by jitsedesmet** [comment]
When merging this PR we should also open 2 new issuees.
1. Is arithmetic widening the right choice? This PR creates a function `arithmetic widening` this function copy's  previous behavior. We should check if this behavior is spec compliant and add links to that part of the spec. Reason: it could be that short + short \ne short. Example: is 240 + 100 still a short? It can not be represented as a short. Should we check this? Or do we need to return a short with value 340 which is not a valid short value.
2. The requested change by @wschella right here https://github.com/comunica/sparqlee/pull/103#discussion_r687755303 . Should be checked closer. resolving that issue should also provide an explanation on any choices made there.

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:18]: I removed it.

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:6]: What name do you suggest?

File: lib/expressions/Term.ts
Comment (jitsedesmet) [reply lib/expressions/Term.ts:73]: I removed both `type` and `typeUrl` and replaced them by `dataType: string` and in certain Literals this becomes `dataType: LiteralTypes`. Since `LiteralTypes` is a subset of string I think this is a pretty solution.

File: lib/util/Consts.ts
Comment (jitsedesmet) [reply lib/util/Consts.ts:85]: Yes, I was able to remove this function.

File: README.md
Comment (jitsedesmet) [reply README.md:338]: This is done now. I checked all functions.

**Comment by jitsedesmet** [comment]
If we wish to implement https://www.w3.org/TR/xpath-31/#promotion completely, we should make sure we cast the `decimal` to `double` or `float` when promoting `decimal`. I will leave this for now. I think we should open a new issue adding this (I think small) feature. I feel like we better do this in a new issue because we might need a package for this? I'm not sure though

File: lib/functions/OverloadTree.ts
Comment (jitsedesmet) [reply lib/functions/OverloadTree.ts]: This conversation has been resolved by type promotion.

File: lib/functions/OverloadTree.ts
Comment (jitsedesmet) [reply lib/functions/OverloadTree.ts]: I'm still convinced we can't really test this. So unless someone can give an example on when this could create a problem I consider this to be done.

Comment (jitsedesmet) [reply lib/functions/OverloadTree.ts]: I added some documentation so people don't break the system and also a test that first performs substitution and that promotion.

**Review by rubensworks — APPROVED** [review]
This stuff wasn't trivial, so really good work!

Before we merge, can you again double-check if all tests (unit, integration, spec, ...) still pass in Comunica?

Could you also list in the main description of this PR, what issues can be closed?

I'll wait a bit for merging until @wschella gives his 👍

**Comment by jitsedesmet** [comment]
Thank you!

> Before we merge, can you again double-check if all tests (unit, integration, spec, ...) still pass in Comunica?

I checked the overal tests and ran test, integration and spec in the actor-init-sarql package. All tests succeeded.

I will add the closing issues.

**Comment by rubensworks** [comment]
@wschella Merged this one already. Feel free to open issues in case you still see issues with this.

**Pull request merged**
